File: worker/src/scripted/states/br.py
# worker/src/scripted/states/br.py
"""Bihar border-tax SCRIPTED runner (web source, human handover).

New to this repo — state knowledge ported from the old repo's
scripted/border_tax/br.py + _handover_runner.

BR specifics (from the old runner):
  owner_info     Entry District default PATNA (validator). BR's checkpost
                 names do NOT match district names (place names or "NOT
                 APPLICABLE") — old "first_option" strategy; the shared
                 helper's first-option fallback carries it.
  vehicle_info   Vehicle Category set-if-empty; Permit Type only if empty
                 (default TEMPORARY PERMIT); Service Type NOT APPLICABLE,
                 set-if-empty with first-option fallback; NO Distance field.
  tax_info       Tax Mode DAYS (default) / QUARTERLY / YEARLY — visible text
                 first, generic value fallback (DAYS=1, QUARTERLY=5,
                 YEARLY=7); a mode the RC doesn't get aborts with
                 abort_reason="tax_mode_not_offered_for_rc" + the offered
                 list. The old repo's own sources DISAGREE on the Tax
                 From/Upto input type (the handover runner's field notes say
                 type="date", br.py's docstring says datetime-local) —
                 runtime detection in scripted.tax_dates decides on the live
                 field either way; DAYS fills Tax Upto.
"""

from __future__ import annotations

from engine.pipeline import Phase
from lifecycle.status import Status

from ..handover import DEFAULT_NEGATIVE_PATTERNS, HandoverConfig, make_handover_phase
from ..wizard import (
    make_open_portal,
    make_tax_info,
    make_vehicle_info,
    owner_info,
    select_service,
)

open_portal = make_open_portal("BR", "br")

vehicle_info = make_vehicle_info(
    set_category_if_empty=True,
    has_distance=False,
    service_set_if_empty=True,
    service_first_option_fallback=True,
    manual_has_category=True,
    manual_has_distance=False,
    manual_datetime_dates=False,
)

_BR_TAX_MODE_VALUES = {"DAYS": "1", "QUARTERLY": "5", "YEARLY": "7"}

tax_info = make_tax_info(
    state_code="BR",
    tax_mode_values=_BR_TAX_MODE_VALUES,
    list_offered_on_miss=True,
)

HANDOVER_CONFIG = HandoverConfig(
    state_name="Bihar",
    receipt_markers=[
        "government of bihar",
        "checkpost tax e-receipt",
        "receipt no",
        "grand total",
    ],
    negative_patterns=list(DEFAULT_NEGATIVE_PATTERNS),
)

PHASES: list[Phase] = [
    Phase("open_portal", open_portal, enter_status=Status.AI_AGENT_STARTED, max_secs=150),
    Phase("select_service", select_service, max_secs=150),
    Phase("owner_info", owner_info, max_secs=300),
    Phase("vehicle_info", vehicle_info, max_secs=180),
    Phase("tax_info", tax_info, max_secs=180),
    make_handover_phase(HANDOVER_CONFIG),
]

File: api/src/validators/states/br.ts
// api/src/validators/states/br.ts
//
// Bihar border-tax validator. Scope: SCRIPTED path only (web source, human
// handover) — BR is not in the fully-automated set, so run.ts rejects
// path=fullyAutomated for it.
//
// Ported knowledge (old repo br.py / defaults / dates):
//   - Tax Mode: DAYS / QUARTERLY / YEARLY.
//   - BR is a NO_SAME_DAY state (duration offset = duration): duration=1 ->
//     taxUpto = taxFrom + 1.
//   - taxUpto is mandatory ONLY for DAYS; QUARTERLY / YEARLY are computed and
//     locked by the portal (empty string tells the worker to read it back).
//   - Defaults: PATNA entry. BR's checkpost names do NOT match districts
//     (place names or "NOT APPLICABLE"), so entryCheckpoint defaults to "" —
//     the worker falls to the FIRST option (old "first_option" strategy).
//     Permit TEMPORARY PERMIT; Service Type NOT APPLICABLE.
//   - The old repo's own sources disagree on BR's Tax From/Upto input type
//     (date vs datetime-local) — the worker's runtime detection decides on
//     the live field, so taxTime is accepted optionally here.
//   - paymentMethod is NOT pinned (operator picks at the portal). Optional,
//     pass-through uppercase, default NETBANKING.
//   - No Distance field on BR, so none is emitted.

import {
    type StateValidator,
    type ValidationResult,
    type FieldError,
    ok,
    fail,
    str,
    normalizeDate,
    normalizeTaxTime,
    isVehicleNumber,
    normalizeVehicleNumber,
} from "../types";

const VALID_TAX_MODES = new Set(["DAYS", "QUARTERLY", "YEARLY"]);

function addDays(date: string, duration: number): string {
    const [year, month, day] = date.split("-").map(Number);
    const d = new Date(Date.UTC(year, month - 1, day));
    d.setUTCDate(d.getUTCDate() + duration);
    return d.toISOString().slice(0, 10);
}

const DEFAULTS = {
    mobileNumber: "",
    // Standard BR entry; change here if the fleet enters elsewhere.
    entryDistrict: "PATNA",
    // BR checkposts don't match districts — empty means "first option" on the
    // worker.
    entryCheckpoint: "",
    permitType: "TEMPORARY PERMIT",
    permitTypeFallback: "TEMPORARY PERMIT",
    serviceType: "NOT APPLICABLE",
    paymentMethod: "NETBANKING",
} as const;

export const brValidator: StateValidator = {
    code: "BR",
    name: "Bihar",

    validate(raw: Record<string, unknown>): ValidationResult {
        const missing: string[] = [];
        const invalid: FieldError[] = [];

        const vehicleNumber = str(raw.vehicleNumber);
        const duration = Number(raw.duration) || 0;
        const taxMode = str(raw.taxMode)?.toUpperCase();
        const taxFrom = normalizeDate(raw.taxFrom);
        let taxUpto = normalizeDate(raw.taxUpto);
        const paymentMethod = str(raw.paymentMethod)?.toUpperCase();
        const requestId = str(raw.requestId);
        const driverId = str(raw.driverId);
        // Optional 24h "HH:MM"; stamped only if the live field is
        // datetime-local (runtime detection). null = not sent (fine).
        const taxTime = normalizeTaxTime(raw.taxTime);

        if (!vehicleNumber) missing.push("vehicleNumber");
        if (!taxMode) missing.push("taxMode");
        if (!requestId) missing.push("requestId");
        if (!driverId) missing.push("driverId");
        if (!taxFrom) {
            if (str(raw.taxFrom)) {
                invalid.push({
                    field: "taxFrom",
                    reason: "expected YYYY-MM-DD or dd/Mon/yyyy",
                });
            } else {
                missing.push("taxFrom");
            }
        }
        if (taxTime === undefined) {
            invalid.push({
                field: "taxTime",
                reason: 'expected 24h "HH:MM", e.g. "09:30" or "14:00"',
            });
        }

        // taxUpto only matters in DAYS mode; QUARTERLY / YEARLY are
        // portal-locked.
        const needsTaxUpto = taxMode === "DAYS";
        if (needsTaxUpto && !taxUpto && duration >= 1) {
            // BR is NO_SAME_DAY: duration=1 -> taxFrom + 1.
            taxUpto = addDays(taxFrom!, duration);
        }
        if (!taxUpto) {
            taxUpto = taxFrom;
        }

        if (vehicleNumber && !isVehicleNumber(vehicleNumber)) {
            invalid.push({
                field: "vehicleNumber",
                reason: "not a valid Indian registration number",
            });
        }
        if (taxMode && !VALID_TAX_MODES.has(taxMode)) {
            invalid.push({
                field: "taxMode",
                reason: "Bihar supports DAYS, QUARTERLY or YEARLY only",
            });
        }
        if (needsTaxUpto && taxFrom && taxUpto && taxUpto < taxFrom) {
            invalid.push({
                field: "taxUpto",
                reason: "must be on or after taxFrom",
            });
        }

        if (missing.length || invalid.length) {
            return fail("Bihar", missing, invalid);
        }

        return ok({
            state: "BIHAR",
            stateCode: "BR",
            // Informational on the scripted path — the operator picks the real
            // method at the portal.
            paymentMethod: paymentMethod || DEFAULTS.paymentMethod,
            requestId: requestId!,
            driverId: driverId!,
            mobileNumber: str(raw.mobileNumber) ?? DEFAULTS.mobileNumber,
            vehicleNumber: normalizeVehicleNumber(vehicleNumber!),
            taxMode: taxMode!,
            taxFrom: taxFrom!,
            // Stripped for QUARTERLY / YEARLY: empty tells the worker the
            // portal auto-fills (and locks) this field.
            taxUpto: needsTaxUpto ? taxUpto! : "",
            ...(taxTime ? { taxTime } : {}),
            entryDistrict: (
                str(raw.entryDistrict) ?? DEFAULTS.entryDistrict
            ).toUpperCase(),
            entryCheckpoint: (
                str(raw.entryCheckpoint) ?? DEFAULTS.entryCheckpoint
            ).toUpperCase(),
            permitType: str(raw.permitType) ?? DEFAULTS.permitType,
            permitTypeFallback:
                str(raw.permitTypeFallback) ?? DEFAULTS.permitTypeFallback,
            serviceType: str(raw.serviceType) ?? DEFAULTS.serviceType,
            // No distance field on BR.
        });
    },
};
